# Contents

1. Scrape and tokenize democratic statements
2. Scrape and tokenize republican statements
3. Build a model and test
4. Make up some phrases.

5. I wanted to semantically map the embedding layer, but due to a bug in this code I have to do it another way. 

6. - Download the "stances" of different campaign sites, and see if this provides a grade of how dem or rep that person is. 

In [1]:
#import dependencies 
from urllib.request import urlopen
from bs4 import BeautifulSoup
import re
import codecs
import nltk
import numpy as np
import pandas as pd
%pylab inline
nltk.download("punkt")
nltk.download("stopwords")

Populating the interactive namespace from numpy and matplotlib
[nltk_data] Downloading package punkt to
[nltk_data]     /Users/noahkasmanoff/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to
[nltk_data]     /Users/noahkasmanoff/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

# 1. Scraping various sources for democratic and republican statements/stances. 

In [2]:
#import democratic data

#scrape DNC website
url = "https://democrats.org/about/party-platform/"
client = urlopen(url)
page_html = client.read()
page_soup = BeautifulSoup(page_html,'html.parser')


dem_corpus = []
print("len of dem corpus: ", len(dem_corpus))
for tag in page_soup.find_all(['h1', 'p']):
  #  print(len(tag.text))
  #  print(tag.text)
  #  print()
    dem_corpus.append(tag.text)
print("len of dem corpus: ", len(dem_corpus))



#scrape some politics website. Does not use other indexing! 
url = "http://www.ontheissues.org/Democratic_Party.htm"
client = urlopen(url)
page_html = client.read()
page_soup = BeautifulSoup(page_html,'html.parser')


for tag in page_soup.find_all(['li']):
    dem_corpus.append(tag.text)
print("len of dem corpus: ", len(dem_corpus))



#via wikipedia
url = "https://en.wikipedia.org/wiki/Political_positions_of_the_Democratic_Party"

client = urlopen(url)
page_html = client.read()
page_soup = BeautifulSoup(page_html,'html.parser')


for tag in page_soup.find_all(['h1', 'p']):
    dem_corpus.append(tag.text)
print("len of dem corpus: ", len(dem_corpus))



#scrape senate democrats website
url = "https://www.democrats.senate.gov/abetterdeal/higher-wages-and-better-jobs"

client = urlopen(url)
page_html = client.read()
page_soup = BeautifulSoup(page_html,'html.parser')


for tag in page_soup.find_all(['h1', 'p']):
  #  print(len(tag.text))
  #  print(tag.text)
  #  print()
    dem_corpus.append(tag.text)
print("len of dem corpus: ", len(dem_corpus))



#continue scraping senate democrats
url = "https://www.democrats.senate.gov/abetterdeal/lower-the-cost-of-living-for-families"

client = urlopen(url)
page_html = client.read()
page_soup = BeautifulSoup(page_html,'html.parser')

for tag in page_soup.find_all(['h1', 'p']):

    dem_corpus.append(tag.text)
print("len of dem corpus: ", len(dem_corpus))


#and again 

url = "https://www.democrats.senate.gov/abetterdeal/tools-to-succeed-in-the-21st-century"

client = urlopen(url)
page_html = client.read()
page_soup = BeautifulSoup(page_html,'html.parser')

for tag in page_soup.find_all(['h1', 'p']):

    dem_corpus.append(tag.text)
print("len of dem corpus: ", len(dem_corpus))

#combine all elements into a continuous body, now resembling a corpus. 
dem_corpus = " ".join(dem_corpus)


tokenizer = nltk.data.load('tokenizers/punkt/english.pickle')


raw_dem_sentences = tokenizer.tokenize(dem_corpus)

def sentence_to_wordlist(raw):
    clean = re.sub("[^a-zA-Z]"," ", raw)
    words = clean.split()
    return words


dem_sentences = []
for raw_dem_sentence in raw_dem_sentences:
    if len(raw_dem_sentence) > 0:
        dem_sentences.append(sentence_to_wordlist(raw_dem_sentence))
print()     
print("Here's an example datapoint.")      
print(raw_dem_sentences[50])
print(sentence_to_wordlist(raw_dem_sentences[50]))



token_count = sum([len(dem_sentence) for dem_sentence in dem_sentences])
print("The dem policy corpus contains {0:,} tokens".format(token_count))


y_dems = np.ones(shape=np.shape(dem_sentences)) #now the labels. 
print("Done!")

len of dem corpus:  0
len of dem corpus:  282
len of dem corpus:  447
len of dem corpus:  489
len of dem corpus:  504
len of dem corpus:  515
len of dem corpus:  526

Here's an example datapoint.
We believe in protecting civil liberties and guaranteeing civil rights and voting rights, women’s rights and workers’ rights, LGBT rights, and rights for people with disabilities.
['We', 'believe', 'in', 'protecting', 'civil', 'liberties', 'and', 'guaranteeing', 'civil', 'rights', 'and', 'voting', 'rights', 'women', 's', 'rights', 'and', 'workers', 'rights', 'LGBT', 'rights', 'and', 'rights', 'for', 'people', 'with', 'disabilities']
The dem policy corpus contains 37,666 tokens
Done!


In [3]:
#Now the republican data. 

#scrape wikipedia description. 
url = "https://en.wikipedia.org/wiki/Political_positions_of_the_Republican_Party"
client = urlopen(url)
page_html = client.read()
page_soup = BeautifulSoup(page_html,'html.parser')


rep_corpus = []
print("len of rep corpus: ", len(rep_corpus))
for tag in page_soup.find_all(['h1', 'p']):
  #  print(len(tag.text))
  #  print(tag.text)
  #  print()
    rep_corpus.append(tag.text)
print("len of rep corpus: ", len(rep_corpus))

#Scrape GOP website for official platform
#Forbidden, so I copy pasted it all into a text file also in this repo. 

#on the issues website
url = "http://www.ontheissues.org/Republican_Party.htm"

client = urlopen(url)
page_html = client.read()
page_soup = BeautifulSoup(page_html,'html.parser')


for tag in page_soup.find_all(['li']):
  #  print(len(tag.text))
  #  print(tag.text)
  #  print()
    rep_corpus.append(tag.text)
print("len of rep corpus: ", len(rep_corpus))



rep_corpus = " ".join(rep_corpus)
tokenizer = nltk.data.load('tokenizers/punkt/english.pickle')




#open the text file and 
rep_platform_corpus_raw = u""
print("Reading '{0}'...".format('rep_platform.txt'))
with codecs.open('rep_platform.txt', "r", "utf-8") as book_file:
    rep_platform_corpus_raw += book_file.read()

rep_corpus += rep_platform_corpus_raw





raw_rep_sentences = tokenizer.tokenize(rep_corpus)


def sentence_to_wordlist(raw):
    clean = re.sub("[^a-zA-Z]"," ", raw)
    words = clean.split()
    return words


rep_sentences = []
for raw_rep_sentence in raw_rep_sentences:
    if len(raw_rep_sentence) > 0:
        rep_sentences.append(sentence_to_wordlist(raw_rep_sentence))
        
print()
print("Here's an example sentence in rep corpus.")
print(raw_rep_sentences[10])
print(sentence_to_wordlist(raw_rep_sentences[10]))


token_count = sum([len(rep_sentence) for rep_sentence in rep_sentences])
print("The rep policy corpus contains {0:,} tokens".format(token_count))


y_reps = np.zeros(shape=np.shape(rep_sentences))


len of rep corpus:  0
len of rep corpus:  72
len of rep corpus:  280
Reading 'rep_platform.txt'...

Here's an example sentence in rep corpus.
[14]
 Republican party leaders strongly believe that free markets and individual achievement are the primary factors behind economic prosperity.
['Republican', 'party', 'leaders', 'strongly', 'believe', 'that', 'free', 'markets', 'and', 'individual', 'achievement', 'are', 'the', 'primary', 'factors', 'behind', 'economic', 'prosperity']
The rep policy corpus contains 57,193 tokens


# Now it's modelling time 

Reference: 
https://github.com/Hvass-Labs/TensorFlow-Tutorials/blob/master/20_Natural_Language_Processing.ipynb

In [4]:
%matplotlib inline
import matplotlib.pyplot as plt
import tensorflow as tf
import numpy as np
from scipy.spatial.distance import cdist


# from tf.keras.models import Sequential  # This does not work!
from tensorflow.python.keras.models import Sequential
from tensorflow.python.keras.layers import Dense, GRU, Embedding
from tensorflow.python.keras.optimizers import Adam
from tensorflow.python.keras.preprocessing.text import Tokenizer
from tensorflow.python.keras.preprocessing.sequence import pad_sequences

# 2. Tokenizing and splitting these statements into a training/testing format. 

In [5]:
raw_rep_sentence

'A Republican administration will champion an open and free internet based on principles of free expression and universal values and will pursue policies to empower citizens and U.S. companies operating in authoritarian countries to circumvent internet firewalls and gain accurate news and information online.'

To recreate the structure of the imdb dataset, the next plan is to concatenate the dem and rep texts and simultaneously with the labels of 1 or 0, and probably shuffle both. 

In [6]:
X_text = np.concatenate((raw_dem_sentences,raw_rep_sentences))
y = np.concatenate((y_dems,y_reps))

In [7]:
from sklearn.model_selection import train_test_split

X_train_text, X_test_text, y_train, y_test = train_test_split(X_text, y ,shuffle=True, test_size=0.33, random_state=42)

In [8]:
len(X_text)

5326

The training and testing data are now separated and shuffled. Next task is to tokenize these words once again, but now this will be based on an integer value for indexing each word (note this contains no semantic meaning), and once fit you can determine how many unique words to include.

In [9]:
num_words=6000
tokenizer = Tokenizer(num_words=num_words)


Fitting: 
The tokenizer scans through all the text and strips unwanted characters, converts everything to lowercase, and builds a vocabulary of all the unique words. It's ok that it uses the entire dataset, as the embedding is only based on the training data (not sure how accurate this is though should follow this up). 

In [10]:
tokenizer.fit_on_texts(X_text)


In [11]:
if num_words is None:
    num_words = len(tokenizer.word_index)

In [12]:
num_words

6000

In [13]:
#this is ordered by the number of occurences of the word. 
tokenizer.word_index


{'the': 1,
 'and': 2,
 'to': 3,
 'of': 4,
 '\r': 5,
 'in': 6,
 'we': 7,
 'a': 8,
 'for': 9,
 'that': 10,
 'aug': 11,
 '2000': 12,
 '2004': 13,
 'our': 14,
 'will': 15,
 'is': 16,
 'on': 17,
 'with': 18,
 'sep': 19,
 'by': 20,
 'as': 21,
 'their': 22,
 'support': 23,
 'are': 24,
 'not': 25,
 'have': 26,
 'be': 27,
 'it': 28,
 'democrats': 29,
 'from': 30,
 'or': 31,
 'should': 32,
 'more': 33,
 'american': 34,
 'government': 35,
 'all': 36,
 'has': 37,
 'federal': 38,
 'must': 39,
 'health': 40,
 'jul': 41,
 'rights': 42,
 'people': 43,
 '2012': 44,
 'an': 45,
 'who': 46,
 'military': 47,
 'states': 48,
 'public': 49,
 'they': 50,
 'national': 51,
 'new': 52,
 'americans': 53,
 'at': 54,
 'which': 55,
 'no': 56,
 'party': 57,
 'against': 58,
 'its': 59,
 'republican': 60,
 'this': 61,
 'education': 62,
 'those': 63,
 'care': 64,
 'other': 65,
 'tax': 66,
 'believe': 67,
 'protect': 68,
 'state': 69,
 'republicans': 70,
 'jun': 71,
 'also': 72,
 'work': 73,
 'u': 74,
 'congress': 75,
 'r

In [14]:
#now it's time to tokenize all of the training and testing data for future use. 
X_train_tokens = tokenizer.texts_to_sequences(X_train_text)

X_test_tokens = tokenizer.texts_to_sequences(X_test_text)

# Padding and truncating

Not all senetences are created equal (both in meaning and length!) To preserve the same size input property of a neural network, this means padding the shorter phrases, and truncating the longer ones. For padding these zeros (which will be interpreted as meaningless) will be placed at the start of the sentence as the sequential nature should play less of a role. Rather than an abrupt jump to zero which could have an adverse effect on the RNN, it is smarter to go from 0 up. 

In [15]:
num_tokens = [len(tokens) for tokens in X_train_tokens + X_test_tokens]
num_tokens = np.array(num_tokens)

In [16]:
np.mean(num_tokens)
#avg num of tokens in a phrase

18.326511453248216

In [17]:
np.max(num_tokens)
#max num of tokens in a phrase

100

In [18]:
max_tokens = np.mean(num_tokens) + 2 * np.std(num_tokens)
max_tokens = int(max_tokens)
max_tokens

40

In [19]:
np.sum(num_tokens < max_tokens) / len(num_tokens)


0.9474277131055201

So by setting it to mean + 2 stds, this covers about 93% of the dataset. Fine by me!

In [20]:
pad = 'pre' #pad in the beginning, reasoning explained earlier. 

X_train_pad = pad_sequences(X_train_tokens, maxlen=max_tokens,
                            padding=pad, truncating=pad)

X_test_pad = pad_sequences(X_test_tokens, maxlen=max_tokens,
                           padding=pad, truncating=pad)

In [21]:
X_train_pad.shape

(3568, 40)

In [22]:
X_test_pad.shape


(1758, 40)

In [23]:
idx = tokenizer.word_index
inverse_map = dict(zip(idx.values(), idx.keys()))

In [24]:
def tokens_to_string(tokens):
    """Helper function to go back to the text."""
    # Map from tokens back to words.
    words = [inverse_map[token] for token in tokens if token != 0]
    
    # Concatenate all words.
    text = " ".join(words)

    return text

In [25]:
len(tokenizer.word_index)

7801

In [28]:
X_train_text[10]

'The Democratic Party supports equal opportunity for all Americans regardless of sex, age, race, ethnicity, sexual orientation, gender identity, religion, creed, or national origin.'

In [29]:
tokens_to_string(X_train_tokens[10])


'the democratic party supports equal opportunity for all americans regardless of sex age race ethnicity sexual orientation gender identity religion creed or national origin'

It works!

# 3. Creating a Recurrent Neural Network 


Now for the implementation of an RNN to classify this data, using the keras structure. 

In [30]:
model = Sequential()
embedding_size = 10  #dimensions of word2vec (basically.)

#this is essentially a word2vec model but it is a part of the RNN, as it trains simultaneously. 
model.add(Embedding(input_dim=num_words,
                    output_dim=embedding_size,
                    input_length=max_tokens,  
                    name='layer_embedding'))



model.add(GRU(units=16, return_sequences=True))
model.add(GRU(units=8, return_sequences=True))
model.add(GRU(units=4))
model.add(Dense(1, activation='sigmoid'))
optimizer = Adam(lr=1e-3)
model.compile(loss='binary_crossentropy',
              optimizer=optimizer,
              metrics=['accuracy'])

In [31]:
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
layer_embedding (Embedding)  (None, 40, 10)            60000     
_________________________________________________________________
gru_1 (GRU)                  (None, 40, 16)            1296      
_________________________________________________________________
gru_2 (GRU)                  (None, 40, 8)             600       
_________________________________________________________________
gru_3 (GRU)                  (None, 4)                 156       
_________________________________________________________________
dense_1 (Dense)              (None, 1)                 5         
Total params: 62,057
Trainable params: 62,057
Non-trainable params: 0
_________________________________________________________________


In [32]:
model.fit(X_train_pad, y_train,
          validation_split=0.05, epochs=15, batch_size=64)


Train on 3389 samples, validate on 179 samples
Epoch 1/15
3389/3389 [==============================] - 7s 2ms/step - loss: 0.6601 - acc: 0.6247 - val_loss: 0.6525 - val_acc: 0.6313
Epoch 2/15
3389/3389 [==============================] - 3s 1ms/step - loss: 0.6501 - acc: 0.6279 - val_loss: 0.6439 - val_acc: 0.6313
Epoch 3/15
3389/3389 [==============================] - 3s 1ms/step - loss: 0.6196 - acc: 0.6598 - val_loss: 0.6321 - val_acc: 0.6592
Epoch 4/15
3389/3389 [==============================] - 3s 993us/step - loss: 0.5174 - acc: 0.7560 - val_loss: 0.5157 - val_acc: 0.7598
Epoch 5/15
3389/3389 [==============================] - 4s 1ms/step - loss: 0.4027 - acc: 0.8389 - val_loss: 0.4382 - val_acc: 0.8156
Epoch 6/15
3389/3389 [==============================] - 4s 1ms/step - loss: 0.3186 - acc: 0.8879 - val_loss: 0.4002 - val_acc: 0.8603
Epoch 7/15
3389/3389 [==============================] - 4s 1ms/step - loss: 0.2610 - acc: 0.9162 - val_loss: 0.3835 - val_acc: 0.8547
Epoch 8/15
33

In [33]:
result = model.evaluate(X_test_pad, y_test)
print("Accuracy: {0:.2%}".format(result[1]))


1758/1758 [==============================] - 1s 472us/step
Accuracy: 82.03%


# Example of mis-classified text



In [34]:
#In order to show an example of mis-classified text, 
#we first calculate the predicted sentiment for the first 500 texts in the test-set.

y_pred = model.predict(x=X_test_pad[0:500])
y_pred = y_pred.T[0]


#These predicted numbers fall between 0.0 and 1.0. 
#We use a cutoff / threshold and say that all values above 0.5 
#are taken to be 1.0 and all values below 0.5 are taken to be 0.0. 
#This gives us a predicted "class" of either 0.0 or 1.0.


cls_pred = np.array([1.0 if p>0.5 else 0.0 for p in y_pred])
cls_true = np.array(y_test[0:500])


We can then get indices for all the texts that were incorrectly classified by comparing all the "classes" of these two arrays.



In [35]:
len(y_dems)

2031

In [36]:
len(y_reps)

3295

In [ ]:
3295 / (3295 + 2031)

In [37]:
incorrect = np.where(cls_pred != cls_true)
incorrect = incorrect[0]

len(incorrect)  #how many out of 500 were wrong?


93

In [38]:
idx = incorrect[-1]


In [39]:
text = X_test_text[idx]
text

'Building the Future: Technology  (Top)\n\nThe digital revolution has transformed how we work, learn, sell, shop, socialize — in short, how we live.'

In [40]:
y_pred[idx]


0.9729476

In [41]:
for val in incorrect:
    print(val)
    idx = val
    text = X_test_text[idx]
    print()
    print(text)
    print("Incorrectly classified as ",y_pred[idx])
    print()

0

As such, they pushed the ADA Amendments Act of 2008, a legal expansion that became law.
Incorrectly classified as  0.023068

22

The Democratic Party is extreme on abortion.
Incorrectly classified as  0.972077

23

They practice boots-on-the-ground conservation in their states every day.
Incorrectly classified as  0.946593

30

Africa: The Promise and the Challenge  (Top)

We recognize Africa’s extraordinary potential.
Incorrectly classified as  0.952158

31

American businesses have now added 14.8 million jobs since private-sector job growth turned positive in early 2010.
Incorrectly classified as  0.0792379

33

To protect religious liberty we will ensure that faith-based institutions, especially those that are vital parts of underserved neighborhoods, do not face discrimination by government.
Incorrectly classified as  0.970309

38

(Jan 2012)
    Cut taxes to stimulate economy and help families.
Incorrectly classified as  0.883378

52

(Aug 2012)
    Workers will have choice to 

# New Data

Now let's try and classify new statements. Some will be obvious, while others should be consfusing.  

In [42]:
text1 = "Climate change is the most important issue to address."
text2 = "We need a safer form of gun control."
text3 = "Obamacare is bad"
text4 = "Our welfare system is broken"
text5 = "The wealthy need to pay their fair share in taxes."
text6 = "Roe v Wade must be overturned."
text7 = "Make America Great Again"
text8 =  "Israel is our greatest ally."  #inherently ambiguous (I hope!)
text9 = "I love russia."
texts = [text1, text2, text3, text4, text5, text6, text7, text8,text9]

In [43]:
tokens = tokenizer.texts_to_sequences(texts)


In [44]:

tokens_pad = pad_sequences(tokens, maxlen=max_tokens,
                           padding=pad, truncating=pad)
tokens_pad.shape

(9, 40)

In [45]:
model.predict(tokens_pad)


array([[ 0.03666094],
       [ 0.96099228],
       [ 0.02572565],
       [ 0.08876403],
       [ 0.9571901 ],
       [ 0.02362171],
       [ 0.95581186],
       [ 0.48675978],
       [ 0.03998385]], dtype=float32)

# Now recreating that prediction method with a raw input for a more direct approach.

In [77]:
text = [input()]
tokens = tokenizer.texts_to_sequences(text)
tokens_pad = pad_sequences(tokens, maxlen=max_tokens,
                           padding=pad, truncating=pad)

x = model.predict(tokens_pad)
if x[0][0] < .5:
    print("Republican")
else:
    print("Democrat")

obamacare
Republican


In [ ]:
if x[0][0] < .5:
    print("Republican")
else:
    print("Democrat")

Qualitatively speaking, this model gueses 6 out of 7, not bad! The democratic ones that serve as the first two are the one's misclassified, and maybe this says something about their policy...

# Semantic Exploration


One perk of the embedding layer is that during the training period, it semantically maps the meaning of different words based on their classification, relative place in the sequence, etc. Certain terms will end up being more meaningful ot the message, while others are not (like stop words, 'and', the',etc.)


By retreiving the embedding layer, we can explore these embeddings and have a little peek under the hood as to what this model has learned about politics. 

In [63]:
#get the layer from the model
layer_embedding = model.get_layer('layer_embedding')


In [64]:
#get the weights from that layer. 
weights_embedding = layer_embedding.get_weights()[0]
print(weights_embedding.shape)

(6000, 10)


In [65]:
for i, vec, in enumerate(weights_embedding):
    #convert to word
    tokenizer.word_index
  

Previously we said that the vec would have 8 dimensions, so this is consistent with those findings here. 

In [66]:
#ex what is the 
token_good = tokenizer.word_index['good']
token_good

238

In [67]:
token_great = tokenizer.word_index['great']
token_great

510

In [68]:
weights_embedding[token_good]


array([ 0.4825162 ,  0.3311922 ,  0.64928097,  0.87284219,  0.24644057,
        0.76803237,  0.47215921,  0.42661455,  0.27512088,  0.05215252], dtype=float32)

In [69]:
weights_embedding[token_great]


array([ 0.49037293,  0.92012733,  0.5688715 ,  0.78711706,  0.28240907,
        0.60960138,  0.57438099,  0.47182074,  0.77179307,  0.7913934 ], dtype=float32)

Since these words are similar, we'd expect their embeddings to be close (speaking from displacement meaning) We see that in some of these dimensions, that is the case. 

# Sorting

Using this helper function, find the distances between every word and sort.



# There is an issue with this function where the index goes beyond what I have originally set. Seems this won't work (yet). 

In [70]:
def print_sorted_words(word, metric='cosine'):
    """
    Print the words in the vocabulary sorted according to their
    embedding-distance to the given word.
    Different metrics can be used, e.g. 'cosine' or 'euclidean'.
    """

    # Get the token (i.e. integer ID) for the given word.
    token = tokenizer.word_index[word]

    # Get the embedding for the given word. Note that the
    # embedding-weight-matrix is indexed by the word-tokens
    # which are integer IDs.
    embedding = weights_embedding[token]

    # Calculate the distance between the embeddings for
    # this word and all other words in the vocabulary.
    distances = cdist(weights_embedding, [embedding],
                      metric=metric).T[0]
    
    # Get an index sorted according to the embedding-distances.
    # These are the tokens (integer IDs) for words in the vocabulary.
    sorted_index = np.argsort(distances)
    
    # Sort the embedding-distances.
    sorted_distances = distances[sorted_index]
    
    # Sort all the words in the vocabulary according to their
    # embedding-distance. This is a bit excessive because we
    # will only print the top and bottom words.
    sorted_words = [inverse_map[token] for token in sorted_index
                  if token != 0]
    # Helper-function for printing words and embedding-distances.
    def _print_words(words, distances):
        for word, distance in zip(words, distances):
            print("{0:.3f} - {1}".format(distance, word))

    # Number of words to print from the top and bottom of the list.
    k = 10

    print("Distance from '{0}':".format(word))

    # Print the words with smallest embedding-distance.
    _print_words(sorted_words[1:k+1], sorted_distances[1:k+1])

    print("...")

    # Print the words with highest embedding-distance.
    _print_words(sorted_words[-k:], sorted_distances[-k:])

In [71]:
print_sorted_words('guns', metric='cosine')


Distance from 'guns':
0.014 - proposed
0.026 - injustice
0.026 - affect
0.026 - crumbling
0.028 - 200
0.029 - speculation
0.032 - permanent
0.035 - congestion
0.036 - races
0.036 - seriousness
...
0.531 - expensive
0.533 - spreads
0.539 - criminal
0.542 - prices
0.549 - oil
0.580 - venture
0.582 - jun
0.596 - sized
0.602 - prominently
0.658 - choices


# New Idea: I've downloaded Josh Gottheimer's stances on the issues, now tokenize and model this to predict each statement, looking at the strongest ones on each as proof.

In [72]:
#open the text file and 
gottheimer_raw = u""
print("Reading '{0}'...".format('gottissues.txt'))
with codecs.open('gottissues.txt', "r", "utf-8") as book_file:
    gottheimer_raw += book_file.read()




Reading 'gottissues.txt'...


In [73]:
gottheimer_raw  #now convert !

'WORKING FOR YOU!\nIn Congress, I’m laser-focused on getting things done for our families and for our children. When I got there, I quickly noticed that the extreme partisanship and gridlock were even worse than any of us could have imagined. However, I also found that from day one, if you actually made an effort to reach across the aisle, to find common ground, you could actually get things done that are good for the country and for New Jersey.\n\nEarly last year, I was elected Co-Chair of the bipartisan Problem Solvers Caucus. It’s a group of 24 Democrats and 24 Republicans who meet every week to see where we can make progress on our toughest issues, from lowering taxes, to creating jobs, to fixing our infrastructure, to balancing the budget. Just like when I was at Ford Motor Company and Microsoft, my goal is to bring people together to get something done. I was honored to be recognized as the most bipartisan freshman Democrat in Congress. For me, it’s not about a partisan agenda; i

In [84]:
gottheimer_text = gottheimer_raw.split('.')

From what I can tell, this data needs to already be in the corpus prior to testing. This means re-running the model every time you grab one of these text files. I need a robust scraper that immediately gets all the text including this one, but keeping them in separate directories until testing later. This is especially because this prediction data is not labelled. 

All the blank spaces correspond to words not in original corpus... 

In [86]:
gottheimer_tokens = tokenizer.texts_to_sequences(gottheimer_text)

In [92]:
gottheimer_tokens_pad = pad_sequences(gottheimer_tokens, maxlen=max_tokens,
                            padding=pad, truncating=pad)

In [95]:
results = model.predict(gottheimer_tokens_pad)

In [96]:
x = 0
for result in results:
    if result > .5:
        x+=1
    

In [97]:
x

105

105 of the len(results)= 135 statements are democratic, meaning this is a democratic candidate (probably )

# Now to repeat this testing this with a republican candidate, for reproducibility, I'll stick with Gott's competitor, John Mcann


In [101]:
#open the text file and 
mcann_raw = u""
print("Reading '{0}'...".format('mcissues.txt'))
with codecs.open('mcissues.txt', "r", "utf-8") as book_file:
    mcann_raw += book_file.read()


mcann_text = mcann_raw.split('.')

Reading 'mcissues.txt'...


In [103]:
mcann_tokens = tokenizer.texts_to_sequences(mcann_text)

In [104]:
mcann_tokens_pad = pad_sequences(mcann_tokens, maxlen=max_tokens,
                            padding=pad, truncating=pad)

In [109]:
results = model.predict(mcann_tokens_pad)

In [122]:
x = 0
for result in results:
    if result > .5:
        x+=1
    

In [123]:
print(x)

61


In [124]:
len(results)

107

In [125]:
mcann_text

['Too many Americans are unemployed, underemployed or working undesirable jobs',
 '\n\nThat’s why I’m running for Congress',
 '\n\nTaxes\n\nI will work to make the tax code more favorable to small businesses which are the backbone of the nation’s economy',
 ' It’s time that Congress put Main Street before Wall Street',
 '\n\nSince historic tax cuts passed, New Jersey residents are concerned about the elimination of the SALT (state and local tax) deduction',
 ' Elect me to Congress and I will work to fix this pertinent concern',
 '\n\nTrade\n\nWe’ve all heard the arguments for free and fair trade',
 ' Like the President, I am for America First trade',
 ' We need to reevaluate trade deals to see what benefits America first',
 '\n\nWe’ve been taken advantage of too often on the world stage—while your jobs have been shipped off to other countries',
 '\n\nAs your Congressman I will never vote for a bill that guts American jobs and creates a negative trade balance',
 ' All trade deals need t

# Conclusion 


By looking at the issues page of a republican and democrat running for the same seat, we see that of the 135 detected statements on the dem's page, 105 were considered dem, good for 78%. On the contrary, the republican's issue page has 107 detected statements, where 61 register as democrat or 57%. At face value this means that the model successfully distinguished which candidate is more democratic of the two, there is a lot more work to be done. For one the training data isn't necessarily perfect and could be drawn from more sources, the strengh of such statements used here may have an impact, among other things. Overall this seems like a good starting point and this project seems to have some potential. 